### PR TITLE
Outputs added directories as they are traversed.

### PR DESCRIPTION
**Pull request by noffle**
This also fixes a bug where adding with -n would prevent directories from
being outputted. Addresses https://github.com/ipfs/go-ipfs/issues/2530.

Bonus: directories no longer appear at the end of the list for `ipfs add` --
they're in order! :dancers:

### CODE REVIEW

**Comment by noffle** [comment]
(I still have some tests to fix up, now that the add order has changed.)

**Comment by whyrusleeping** [comment]
No. The change to output directories at the end of the add was a conscious decision, files in a directory may be added before or after the directory itself is 'added' so the hash of the directory may change. No order is guaranteed about the stream of objects being added in an add

**Comment by noffle** [comment]
Ah, this makes sense now. Consider a case like:

```
$ ipfs add foo/bar foo/baz
```

We need to create the `foo` dir when `foo/bar` is created, but we can't output its hash yet, since after `foo/baz` is added it will have changed -- so directories definitely need to have their hash computed last.

However, this doesn't mean we must **output** them at the end: we can still perform a formatting step post-add that will provide a more traditional output structure (subfiles THEN containing dir).

**Comment by noffle** [comment]
I ended up modifying the existing logic to use MFS for traversal instead of relying on the dag service directly. I had to tweak how `mfs.Directory` was doing caching to ensure that the children are available after being added by `AddNode`, but otherwise this went pretty smoothly after I grokked the code structure.

File: core/coreunix/add.go [hunk 1]
```diff
@@ -173,48 +173,63 @@ func (adder *Adder) PinRoot() error {
...
 		return nil, err
 	}
 
 	var name string
 	if !adder.Wrap {
-		name = root.Links[0].Name
-		child, err := root.Links[0].GetNode(adder.ctx, adder.node.DAG)
+		name = rootNode.Links[0].Name
+
+		root, err = adder.mr.GetValue().(*mfs.Directory).Child(name)
```
Comment (whyrusleeping) [inline core/coreunix/add.go]: we'll want to do some type checking here:
```
dir, ok := adder.mr.GetValue().(*mfs.Directory)
if !ok {
   // errrorrrr
}

root, err = dir.Child(name)
...
```

File: core/coreunix/add.go [hunk 2]
```diff
@@ -173,48 +173,63 @@ func (adder *Adder) PinRoot() error {
...
-		root = child
 	}
 
 	err = adder.outputDirs(name, root)
 	if err != nil {
 		return nil, err
 	}
 
-	err = adder.mr.Close()
+	if write {
```
Comment (whyrusleeping) [inline core/coreunix/add.go]: why is this close conditionaled out?

File: core/coreunix/add.go [hunk 3]
```diff
@@ -173,48 +173,63 @@ func (adder *Adder) PinRoot() error {
...
+		return err
+	}
+
+	if !bytes.Equal(nd.Data, folderData) || fs.Type() != mfs.TDir {
 		return nil
 	}
 
-	for _, l := range nd.Links {
-		child, err := l.GetNode(adder.ctx, adder.node.DAG)
+	dir := fs.(*mfs.Directory)
```
Comment (whyrusleeping) [inline core/coreunix/add.go]:
```
dir, ok := fs.(*mfs.Directory)
if !ok {
   ............
}
```

File: mfs/ops.go [hunk 4]
```diff
@@ -100,9 +100,9 @@ func PutNode(r *Root, path string, nd *dag.Node) error {
 
 // Mkdir creates a directory at 'path' under the directory 'd', creating
 // intermediary directories as needed if 'mkparents' is set to true
-func Mkdir(r *Root, pth string, mkparents bool, flush bool) error {
+func Mkdir(r *Root, pth string, mkparents bool, flush bool) (*Directory, error) {
 	if pth == "" {
-		return nil
+		return nil, nil
```
Comment (whyrusleeping) [inline mfs/ops.go]: returning nil and nil feels weird to me... code might assume that no error means you get back a valid directory.

File: test/sharness/t0040-add-and-cat.sh [hunk 5]
```diff
@@ -285,6 +285,23 @@ test_expect_success "'ipfs add -r' output looks good" '
 	test_cmp expected actual
 '
 
+test_expect_success "'ipfs add -rn' succeeds" '
+	mkdir mountdir/moons &&
+	echo "Hello Europa!" >mountdir/moons/europa.txt &&
+	echo "Hello Titan!" >mountdir/moons/titan.txt &&
```
Comment (whyrusleeping) [inline test/sharness/t0040-add-and-cat.sh]: I'd like to see another deeper directory tested here, just to be sure that we don't just have the root dir special cased for printing.

File: core/coreunix/add.go (hunk 2 above)
Comment (noffle) [inline core/coreunix/add.go]: It looks like `Close` triggers an mfs update signal. Isn't this undesirable?

Comment (whyrusleeping) [inline core/coreunix/add.go]: no. We want to ensure (from a top level perspective) that the mfs structure is properly cleaned up, it triggers an update if something registered for updates, but this instance of it doesnt.

Comment (noffle) [inline core/coreunix/add.go]: ACK

**Comment by noffle** [comment]
:tada: Ready for merge!

File: core/coreunix/add.go [hunk 6]
```diff
@@ -200,21 +207,35 @@ func (adder *Adder) Finalize() (*dag.Node, error) {
 		return nil, err
 	}
 
-	return root, nil
+	rootNode, err = root.GetNode()
```
Comment (whyrusleeping) [inline core/coreunix/add.go]: can actually just do `return root.GetNode()` its a little cleaner looking

File: core/coreunix/add.go [hunk 7]
```diff
@@ -200,21 +207,35 @@ func (adder *Adder) Finalize() (*dag.Node, error) {
...
 
-func (adder *Adder) outputDirs(path string, nd *dag.Node) error {
-	if !bytes.Equal(nd.Data, folderData) {
+func (adder *Adder) outputDirs(path string, fs mfs.FSNode) error {
+	nd, err := fs.GetNode()
+	if err != nil {
+		return err
+	}
+
+	if !bytes.Equal(nd.Data, folderData) || fs.Type() != mfs.TDir {
```
Comment (whyrusleeping) [inline core/coreunix/add.go]: since we're using the mfs nodes for this, we can skip the ugly bytewise protobuf comparison.

File: mfs/ops.go [hunk 8]
```diff
@@ -100,9 +100,9 @@ func PutNode(r *Root, path string, nd *dag.Node) error {
 
 // Mkdir creates a directory at 'path' under the directory 'd', creating
 // intermediary directories as needed if 'mkparents' is set to true
-func Mkdir(r *Root, pth string, mkparents bool, flush bool) error {
+func Mkdir(r *Root, pth string, mkparents bool, flush bool) (*Directory, error) {
```
Comment (whyrusleeping) [inline mfs/ops.go]: why the change to making Mkdir return the directory? we don't appear to actually use that anywhere

Comment (noffle) [inline mfs/ops.go]: You're right: this was left over from an earlier approach I took. It seems nice to have (giving back the dir you asked me to create), but I can nix it if you'd rather not keep it around.

Comment (whyrusleeping) [inline mfs/ops.go]: until we have a solid usecase for it, lets not change the existing API

File: core/coreunix/add.go (hunk 7 above)
Comment (noffle) [inline core/coreunix/add.go]: I noticed that tests break when I remove it -- but only the trickledag ones. Perhaps the root node created isn't being set to a `TDir`?

File: mfs/ops.go (hunk 8 above)
Comment (noffle) [inline mfs/ops.go]: Sure thing -- reverted!

File: core/coreunix/add.go (hunk 6 above)
Comment (noffle) [inline core/coreunix/add.go]: Agreed -- changed.

File: core/coreunix/add.go (hunk 7 above)
Comment (whyrusleeping) [inline core/coreunix/add.go]: hrm... i'll pull it down and take a look

**Comment by noffle** [comment]
Anything you need done in here, @whyrusleeping?

**Comment by whyrusleeping** [comment]
Nope, this LGTM
